### PR TITLE
tests: adds WSP functional test for local package

### LINKED ISSUE #106: WSP: fix local package crawling
The mechanism of WSP spider that crawls from local stored folders (given a path) doesn't work for paths including files. 

E.g: `/path/to/file.zip` works but `/path/to/folder_with_zip_files/` doesn't.

### PR DESCRIPTION
**Pull request by spirosdelviniotis**
* Adds WSP functional test for local package path.
* Re-factored existing WSP functional test (add setup-teardown fixtures).
* Re-factored `utils.ftp_list_files` to `utils.list_files` for re-usability.
* Fixed WSP local package crawling mechanism.

Closes #106

Signed-off-by: Spiros Delviniotis <spyridon.delviniotis@cern.ch>

### CODE REVIEW

File: hepcrawl/spiders/wsp_spider.py [hunk 1]
```diff
@@ -89,16 +89,24 @@ def __init__(self, package_path=None, ftp_folder="WSP", ftp_host=None, ftp_netrc
     def start_requests(self):
         """List selected folder on remote FTP and yield new zip files."""
         if self.package_path:
-            yield Request(self.package_path, callback=self.handle_package_file)
+            dummy, new_files = local_list_files(
```
Comment (david-caro) [inline hepcrawl/spiders/wsp_spider.py]: In python, for non-used variables like this `dummy` it's 'convention' to use just `_`.

File: hepcrawl/spiders/wsp_spider.py [hunk 2]
```diff
@@ -89,16 +89,24 @@ def __init__(self, package_path=None, ftp_folder="WSP", ftp_host=None, ftp_netrc
...
+                self.package_path,
+                self.target_folder
+            )
+
+            for _file in new_files:
```
Comment (david-caro) [inline hepcrawl/spiders/wsp_spider.py]: Instead of `_file` and `new_files`, it's better to use `file_path` and `new_files_paths` so you know it's a string representing the file path, and not a file object (that is usually clear when you write the code yourself, but for someone else it might become very confusing when debugging).

File: hepcrawl/utils.py [hunk 3]
```diff
@@ -40,30 +44,48 @@ def unzip_xml_files(filename, target_folder):
...
-            destination_file = os.path.join(target_folder, filename)
-            source_file = os.path.join(server_folder, filename)
-            if not os.path.exists(destination_file):
-                missing_files.append(source_file)
-            all_files.append(source_file)
+        return list_files(server_folder, target_folder, files)
+
+
+def local_list_files(local_folder, target_folder):
+    """List files from given package folder to target folder."""
```
Comment (david-caro) [inline hepcrawl/utils.py]: I don't understand what this function is supposed to do by the comment, can you rephrase? What does it mean that the files 'are listed from a dir to another dir'?

Comment (david-caro) [inline hepcrawl/utils.py]: I know it was not clear before either, but better not add more confusion ;)

File: hepcrawl/utils.py [hunk 4]
```diff
@@ -40,30 +44,48 @@ def unzip_xml_files(filename, target_folder):
...
+    files = os.listdir(local_folder)
+    return list_files(local_folder, target_folder, files)
+
+
+def list_files(remote_folder, target_folder, files):
```
Comment (david-caro) [inline hepcrawl/utils.py]: Maybe this should be called `list_missing_files` instead?
The `files` parameter would be better called `file_names`.
It feels to me that this is doing too many things at the same time... can you think of a way to split getting the list of missing files, with adding the paths to the files? (so there's no need to return both `all_files` and `missing_files`, making it easier to understand, test and use)

Comment (david-caro) [inline hepcrawl/utils.py]: Actually, the `all_files` parameter does not seem to be used anywhere... we can remove it, then there's no need for a `dummy` variable anymore.

File: tests/functional/WSP/test_wsp.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        'CRAWLER_ARGUMENTS': {
+            'package_path': package_location,
+        }
+    }
+
+    cleaner()
+    cleaner(package_location + 'IDAQPv20i01-03160015-1510863')
+
+
+def cleaner(path='/tmp/WSP/'):
```
Comment (david-caro) [inline tests/functional/WSP/test_wsp.py]: This is a bad name for a function, cleaner gives the impression to be an object, that you call on (it's an entity that has meaning by itself) while this is actually a function, maybe rename to something like `clean_dir`?

**Comment by david-caro** [comment]
Can you also try to separate them in individual commits? (it's a bit painful to do now, so next time try to create a lot of commits instead, that can be squashed easily later ;) )

**Comment by david-caro** [comment]
btw. ran locally the functional tests with this pr, and got error (checking if it's my issue):

```
____________________________________________________ test_wsp_ftp _____________________________________________________

set_up_ftp_environment = {'CRAWLER_ARGUMENTS': {'ftp_host': 'ftp_server', 'ftp_netrc': '/code/tests/functional/WSP/fixtures/ftp_server/.netrc'}, 'CRAWLER_HOST_URL': 'http://scrapyd:6800', 'CRAWLER_PROJECT': 'hepcrawl'}
expected_results = [{'abstracts': [{'source': 'WSP', 'value': 'Abstract L\xe9vy bla-bla bla blaaa blaa bla blaaa blaa, bla blaaa blaa. Bla b..., City, City_code 123456, C. R. Country_1'}], 'full_name': 'author_surname_2, author_name_2'}], 'citeable': True, ...}]

    def test_wsp_ftp(set_up_ftp_environment, expected_results):
        crawler = get_crawler_instance(set_up_ftp_environment.get('CRAWLER_HOST_URL'))
    
        # The test must wait until the docker environment is up (takes about 5 seconds).
        sleep(5)
    
        results = CeleryMonitor.do_crawl(
            app=app,
            monitor_timeout=5,
            monitor_iter_limit=100,
            crawler_instance=crawler,
            project=set_up_ftp_environment.get('CRAWLER_PROJECT'),
            spider='WSP',
            settings={},
>           **set_up_ftp_environment.get('CRAWLER_ARGUMENTS')
        )

tests/functional/WSP/test_wsp.py:110: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
hepcrawl/testlib/celery_monitor.py:81: in do_crawl
    **crawler_arguments
/hepcrawl_venv/lib/python2.7/site-packages/scrapyd_api/wrapper.py:184: in schedule
    json = self.client.post(url, data=data)
/hepcrawl_venv/lib/python2.7/site-packages/requests/sessions.py:535: in post
    return self.request('POST', url, data=data, json=json, **kwargs)
/hepcrawl_venv/lib/python2.7/site-packages/scrapyd_api/client.py:38: in request
    return self._handle_response(response)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <scrapyd_api.client.Client object at 0x4c685d0>, response = <Response [200]>

    def _handle_response(self, response):
        """
            Handles the response received from Scrapyd.
            """
        if not response.ok:
            raise ScrapydResponseError(
                "Scrapyd returned a {0} error: {1}".format(
                    response.status_code,
                    response.text))
    
        try:
            json = response.json()
        except ValueError:
            raise ScrapydResponseError("Scrapyd returned an invalid JSON "
                                       "response: {0}".format(response.text))
        if json['status'] == 'ok':
            json.pop('status')
            return json
        elif json['status'] == 'error':
>           raise ScrapydResponseError(json['message'])
E           ScrapydResponseError: spider 'WSP' not found
```

File: hepcrawl/utils.py [hunk 6]
```diff
@@ -66,16 +66,25 @@ def ftp_list_files(server_folder, target_folder, server, user, password, passive
...
+
+def local_list_files(local_folder, target_folder):
+    """List files from given local path to target folder."""
+    file_names = os.listdir(local_folder)
+    return list_missing_files(local_folder, target_folder, file_names)
+
+
+def list_missing_files(remote_folder, target_folder, file_names):
+    missing_files = []
+    for filename in file_names:
```
Comment (david-caro) [inline hepcrawl/utils.py]: use the same style for `filename` and `file_names`

File: tests/functional/WSP/test_wsp.py [hunk 7]
```diff
@@ -50,14 +51,14 @@ def expected_results():
...
 
-@pytest.fixture(scope="module")
-def set_up_environment():
+@pytest.fixture(scope="function")
+def set_up_ftp_environment():
     netrc_location = os.path.join(os.path.dirname(
         os.path.realpath(__file__)),
-        'fixtures/ftp_server/.netrc'
+        os.path.join('fixtures', 'ftp_server', '.netrc')
     )
```
Comment (david-caro) [inline tests/functional/WSP/test_wsp.py]: The indent here is messed up

File: tests/functional/WSP/test_wsp.py [hunk 8]
```diff
@@ -66,9 +67,34 @@ def set_up_environment():
...
+    yield {
+        'CRAWLER_HOST_URL': 'http://scrapyd:6800',
+        'CRAWLER_PROJECT': 'hepcrawl',
+        'CRAWLER_ARGUMENTS': {
+            'package_path': package_location,
+        }
+    }
+
+    clean_dir()
+    clean_dir(package_location + 'IDAQPv20i01-03160015-1510863')
```
Comment (david-caro) [inline tests/functional/WSP/test_wsp.py]: Isn't this missing some `os.path.join`?

File: tests/functional/WSP/test_wsp.py [hunk 9]
```diff
@@ -66,9 +67,34 @@ def set_up_environment():
...
     }
 
+    clean_dir()
+
+
+@pytest.fixture(scope="function")
+def set_up_local_environment():
+    package_location = os.path.join(os.path.dirname(
+        os.path.realpath(__file__)),
+        'fixtures/ftp_server/WSP/'
```
Comment (david-caro) [inline tests/functional/WSP/test_wsp.py]: bad indent

File: tests/functional/WSP/test_wsp.py (hunk 8 above)
Comment (david-caro) [inline tests/functional/WSP/test_wsp.py]:
```python
for _, dirs, files  in next(os.walk(package_location)):
    for dir_name in dirs:
        clean_dir(dir_name)
    for file_name in files:
        os.unlink(file_name)
```

File: tests/functional/WSP/test_wsp.py [hunk 10]
```diff
@@ -78,10 +104,28 @@ def test_wsp_normal_set_of_records(set_up_environment, expected_results):
...
         crawler_instance=crawler,
-        project=set_up_environment.get('CRAWLER_PROJECT'),
+        project=set_up_ftp_environment.get('CRAWLER_PROJECT'),
+        spider='WSP',
+        settings={},
+        **set_up_ftp_environment.get('CRAWLER_ARGUMENTS')
+    )
+
+    assert [override_generated_fields(result) for result in results] == \
+           [override_generated_fields(expected) for expected in expected_results]
```
Comment (david-caro) [inline tests/functional/WSP/test_wsp.py]: Save the lists in temp variables, then just compare them.

File: tests/functional/WSP/test_wsp.py [hunk 11]
```diff
@@ -78,10 +104,28 @@ def test_wsp_normal_set_of_records(set_up_environment, expected_results):
...
+        monitor_iter_limit=100,
+        crawler_instance=crawler,
+        project=set_up_local_environment.get('CRAWLER_PROJECT'),
         spider='WSP',
         settings={},
-        **set_up_environment.get('CRAWLER_ARGUMENTS')
+        **set_up_local_environment.get('CRAWLER_ARGUMENTS')
     )
 
     gottern_results = [override_generated_fields(result) for result in results]
```
Comment (david-caro) [inline tests/functional/WSP/test_wsp.py]: s/gottern/gotten/

File: tests/functional/WSP/test_wsp.py [hunk 12]
```diff
@@ -86,7 +86,16 @@ def set_up_local_environment():
...
-    clean_dir(package_location + 'IDAQPv20i01-03160015-1510863')
+    remove_generated_files(package_location)
+
+
+def remove_generated_files(package_location):
+    for _, dirs, files in os.walk(package_location):
+        for dir_name in dirs:
+            clean_dir(os.path.join(package_location, dir_name))
+        for file_name in files:
+            if '.zip' not in file_name:
```
Comment (david-caro) [inline tests/functional/WSP/test_wsp.py]: not file_name.endswith('.zip')

File: tests/functional/WSP/test_wsp.py [hunk 13]
```diff
@@ -86,7 +86,16 @@ def set_up_local_environment():
     }
 
     clean_dir()
-    clean_dir(package_location + 'IDAQPv20i01-03160015-1510863')
+    remove_generated_files(package_location)
+
+
+def remove_generated_files(package_location):
+    for _, dirs, files in os.walk(package_location):
```
Comment (david-caro) [inline tests/functional/WSP/test_wsp.py]: instead of a for, you can just use the first result:
```python
_, dirs, files = next(os.walk(...))
for dir_name in dirs:
...
```

File: hepcrawl/utils.py [hunk 14]
```diff
@@ -66,16 +66,25 @@ def ftp_list_files(server_folder, target_folder, server, user, password, passive
...
-            if not os.path.exists(destination_file):
-                missing_files.append(source_file)
-            all_files.append(source_file)
-    return all_files, missing_files
+        file_names = host.listdir(host.curdir + '/' + server_folder)
+        return list_missing_files(server_folder, target_folder, file_names)
+
+
+def local_list_files(local_folder, target_folder):
+    """List files from given local path to target folder."""
```
Comment (david-caro) [inline hepcrawl/utils.py]: This comment does not make much sense.

File: hepcrawl/utils.py [hunk 15]
```diff
@@ -66,16 +66,24 @@ def ftp_list_files(server_folder, target_folder, server, user, password, passive
...
-        missing_files = []
-        all_files = []
-        for filename in files:
-            destination_file = os.path.join(target_folder, filename)
-            source_file = os.path.join(server_folder, filename)
-            if not os.path.exists(destination_file):
-                missing_files.append(source_file)
-            all_files.append(source_file)
-    return all_files, missing_files
+        file_names = host.listdir(host.curdir + '/' + server_folder)
```
Comment (david-caro) [inline hepcrawl/utils.py]: os.path.join

**Comment by spirosdelviniotis** [comment]
Ready! :dancer:
